include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,328 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef __NBL_ASSET_C_NORMAL_MAP_TO_DERIVATIVE_FILTER_H_INCLUDED__
+#define __NBL_ASSET_C_NORMAL_MAP_TO_DERIVATIVE_FILTER_H_INCLUDED__
+
+#include "nbl/core/core.h"
+
+#include <type_traits>
+#include <functional>
+ 
+#include "nbl/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "nbl/asset/filters/CSwizzleAndConvertImageFilter.h"
+#include "CConvertFormatImageFilter.h"
+
+namespace nbl
+{
+namespace asset
+{
+
+template<typename Swizzle>
+class CNormalMapToDerivativeFilterBase : public impl::CSwizzleableAndDitherableFilterBase<false, false, Swizzle, IdentityDither>
+{
+	public:
+		class CNormalMapToDerivativeStateBase : public impl::CSwizzleableAndDitherableFilterBase<false, false, Swizzle, IdentityDither>::state_type
+		{
+			public:
+
+				using decodeType = float;
+				static inline constexpr size_t decodeTypeByteSize = sizeof(float);
+				static inline constexpr size_t forcedScratchChannelAmount = 2;
+				uint8_t*	scratchMemory = nullptr;										//!< memory covering all regions used for temporary filling within computation of sum values
+				size_t	scratchMemoryByteSize = {};											//!< required byte size for entire scratch memory
+				bool normalizeImageByTotalABSValues = true;									//!< force normalizing by maximum absolute values
+
+				/*
+					layerCount - layer count used to execute the filter, not global layer count!
+					extent - extent of input image at chosen mip map level
+				*/
+
+				static inline size_t getRequiredScratchByteSize(size_t layerCount, asset::VkExtent3D extent)
+				{
+					size_t retval = extent.width * extent.height * extent.depth * decodeTypeByteSize * forcedScratchChannelAmount + (layerCount * decodeTypeByteSize * forcedScratchChannelAmount);
+					
+					return retval;
+				}
+
+				/*
+					Layer ID is relative to outBaseLayer in state
+				*/
+
+				enum E_SCALE_FACTOR
+				{
+					ESF_X,
+					ESF_Y,
+					ESF_COUNT
+				};
+
+				const float getAbsoluteLayerScaleValue(size_t layer, E_SCALE_FACTOR scaleFactor)
+				{
+					if (!scaleValuesPointer)
+					{
+						auto offset = layer * forcedScratchChannelAmount + scaleFactor == ESF_X ? 0 : 1;
+						return *(scaleValuesPointer + offset);
+					}
+					else
+						return 0; // or maybe assert?
+				}
```

Comment:
not sure why you need an enum

you could just write into a `float[2]` or return a `vectorSIMDf`

Reply:
corrected in following commit

Reply:
where is that commit?